internal/statemachine/classic_test.go

```diff
@@ -3741,8 +3769,141 @@ func TestFailedCreateChroot(t *testing.T) {
 	os.RemoveAll(stateMachine.stateMachineFlags.WorkDir)
 }
 
-// TestFailedInstallPackages tests failure cases in installPackages
-func TestFailedInstallPackages(t *testing.T) {
+// TestStateMachine_installPackages_checkcmds checks commands to install packages order is ok
+func TestStateMachine_installPackages_checkcmds(t *testing.T) {
+	asserter := helper.Asserter{T: t}
+	var stateMachine ClassicStateMachine
+	stateMachine.commonFlags, stateMachine.stateMachineFlags = helper.InitCommonOpts()
+	stateMachine.commonFlags.Debug = true
+	stateMachine.parent = &stateMachine
+	stateMachine.commonFlags.OutputDir = "/tmp"
+
+	err := stateMachine.makeTemporaryDirectories()
+	asserter.AssertErrNil(err, true)
+	err = os.MkdirAll(stateMachine.tempDirs.chroot, 0755)
+	asserter.AssertErrNil(err, true)
+
+	t.Cleanup(func() { os.RemoveAll(stateMachine.stateMachineFlags.WorkDir) })
+
+	mockCmder := NewMockExecCommand()
+
+	execCommand = mockCmder.Command
+	t.Cleanup(func() { execCommand = exec.Command })
+
+	stdout, restoreStdout, _ := helper.CaptureStd(&os.Stdout)
+	t.Cleanup(func() { restoreStdout() })
+
+	helperBackupAndCopyResolvConf = mockBackupAndCopyResolvConfSuccess
+	t.Cleanup(func() {
+		helperBackupAndCopyResolvConf = helper.BackupAndCopyResolvConf
+	})
+
+	err = stateMachine.installPackages()
+	asserter.AssertErrNil(err, true)
+
+	restoreStdout()
+	readStdout, _ := io.ReadAll(stdout)
+
+	expectedCmds := []*regexp.Regexp{
+		regexp.MustCompile("^mount -t devtmpfs devtmpfs-build /tmp.*/chroot/dev$"),
+		regexp.MustCompile("^mount -t devpts devpts-build -o nodev,nosuid /tmp.*/chroot/dev/pts$"),
+		regexp.MustCompile("^mount -t proc proc-build /tmp.*/chroot/proc$"),
+		regexp.MustCompile("^mount -t sysfs sysfs-build /tmp.*/chroot/sys$"),
+		regexp.MustCompile("^mount --bind .*/scratch/run.* .*/chroot/run$"),
+		regexp.MustCompile("^chroot /tmp.*/chroot apt update$"),
+		regexp.MustCompile("^chroot /tmp.*/chroot apt install --assume-yes --quiet --option=Dpkg::options::=--force-unsafe-io --option=Dpkg::Options::=--force-confold$"),
+		regexp.MustCompile("^udevadm settle$"),
+		regexp.MustCompile("^mount --make-rprivate /tmp.*/chroot/run$"),
+		regexp.MustCompile("^umount --recursive /tmp.*/chroot/run$"),
+		regexp.MustCompile("^mount --make-rprivate /tmp.*/chroot/sys$"),
+		regexp.MustCompile("^umount --recursive /tmp.*/chroot/sys$"),
+		regexp.MustCompile("^mount --make-rprivate /tmp.*/chroot/proc$"),
+		regexp.MustCompile("^umount --recursive /tmp.*/chroot/proc$"),
+		regexp.MustCompile("^mount --make-rprivate /tmp.*/chroot/dev/pts$"),
+		regexp.MustCompile("^umount --recursive /tmp.*/chroot/dev/pts$"),
+		regexp.MustCompile("^mount --make-rprivate /tmp.*/chroot/dev$"),
+		regexp.MustCompile("^umount --recursive /tmp.*/chroot/dev$"),
+	}
+
+	gotCmds := strings.Split(strings.TrimSpace(string(readStdout)), "\n")
+	if len(expectedCmds) != len(gotCmds) {
+		t.Fatalf("%v commands to be executed, expected %v", len(gotCmds), len(expectedCmds))
```

Comment:
Bonus points for printing out what commands have been executed, so we can instantly see what got broken.